### PR TITLE
[ENG-3898][ENG-3899]Model support for OutcomeArtifact update and delete

### PR DESCRIPTION
**Pull request by jwalz**
## Purpose
Nuances of updating an OutcomeArtifact's linked Identifier and of deleting an OutcomeArtifact ought to live on the model. This PR implements that logic along with the logging of changes to OutcomeArtifact PIDs on the primary Registration.

## Changes
* Add new fields `finalized` and `deleted` to the OutcomeArtifact
  * `finalized` controls if an OutcomeArtifact can be deleted from the database and which actions from the API get logged
  * `deleted` marks OutcomeArtifacts that have been removed from an Outcome while leaving the database entity intact for referential integrity purposes
* Add `OutcomeArtifact.update_identifier` function
  * Replaces the `identifier` value with one matching the provided pid value and type
  * Attempts to delete the old Identifier to avoid leaving orphaned values in the system
  * If `finalized==True` and an API request is provided, log the update on the parent Outcome
* Add `OutcomeArtifact.finalize` function
  * Sets finalized to True.
  * If an API request is provided log the Artifact as having been added to the parent Outcome
* Add `OutcomeArtifact.delete` function
  *If `finalized==False`, delete the entry from the database
  *If `finalized==True`, just mark as deleted
    * If an API request is provided, log the removal to the parent Outcome
  * Attempts to delete the linked Identifier to avoid leaving orphaned values in the system
* Add `Identifier.delete` function
  * Deletes the identifier if it has no referent and is not referenced by any OutcomeArtifacts
  * Raises IdentifierHasReferencesError if either condition above is not met
* Add `OutcomeAction` Enum for classifying changes to an Outcome
* Add `Outcome.log_artifact_change` function to...log changes to artifacts
* Add new NodeLog actions for Resource changes
* "back-port" the annotations for grabbing `primary_resource_guid` on the OutcomeArtifact to avoid test breakages
* Remove unused OutcomeArtifact functionality to add an OutcomeArtifact by DOI
* Tests
 

## QA Notes
Tests for NodeLogs will be added in API tests

## Ticket
https://openscience.atlassian.net/browse/ENG-3898 
https://openscience.atlassian.net/browse/ENG-3899

### CODE REVIEW

File: osf/models/outcome_artifacts.py [hunk 1]
```diff
@@ -95,3 +98,85 @@ class Meta:
...
+    def finalize(self, api_request=None):
+        '''Sets `finalized` to True and handles callbacks.
+
+        Logs the change on the parent Outcome if invoked via the API.
+
+        Parameters:
+        api_request: The api_request data from the API call that initiated the change.
+        '''
+        if not (self.identifier and self.identifier.value and self.artifact_type):
+            return False
```
Comment (brianjgeiger) [inline osf/models/outcome_artifacts.py:143]: Does this need to return `True` if it succeeds? Or is this just to bail early?

File: osf/models/nodelog.py [hunk 2]
```diff
@@ -143,6 +143,10 @@ class NodeLog(ObjectIDMixin, BaseModel):
 
     MIGRATED_QUICK_FILES = 'migrated_quickfiles'
 
+    RESOURCE_ADDED = 'resource_identifier_added'
+    RESOURCE_UPDATED = 'resource_identifier_udpated'
+    RESOURCE_REMOVED = 'resource_identifier_removed'
```
Comment (brianjgeiger) [inline osf/models/nodelog.py:148]: I wonder if we should also update logActionsList.json and anonymousLogActionsList.json to handle the new log messages. It won't be immediately practical, but if we ever do end up updating the ember front end to display logs, we'll want entries in there or it'll probably be overlooked.

File: osf/models/outcomes.py [hunk 3]
```diff
@@ -72,11 +79,12 @@ class Outcome(ObjectIDMixin, EditableFieldsMixin, BaseModel):
...
-            outcome=self,
-            pid_value=pid,
-            pid_type=pid_type,
-            artifact_type=artifact_type,
-            create_identifier=create_identifier
+    def log_artifact_change(self, action, artifact, api_request, **log_params):
+        nodelog_action = NODE_LOGS_FOR_OUTCOME_ACTION[action]
+        nodelog_params = {'artifact_id': artifact._id, **log_params}
+
+        self.primary_osf_resource.add_log(
```
Comment (mfraezz) [inline osf/models/outcomes.py:86]: We typically don't add new logs to `Registration` objects (the only valid `primary_osf_resource` type for now), and they aren't displayed anywhere. If we add a display for them later, or when `Node` becomes a permissible `primary` resource type, it make cause issues that neither `should_hide` is `True`, nor does `logsActionList.json` have entries for the new logs.

**Review by brianjgeiger — CHANGES_REQUESTED** [review]
One minor question, and maybe a bit of extra work.

File: osf/models/outcomes.py (hunk 3 above)
Comment (jwalz) [reply osf/models/outcomes.py:86]: I'm would need a reeeaalllly good reason to  keep logging Registration-specific activity to the `registered_node`, especially now that we're well into the world of No Project Registrations. That said I'm fine updating `logActionsList.json`.

File: osf/models/outcome_artifacts.py (hunk 1 above)
Comment (jwalz) [reply osf/models/outcome_artifacts.py:143]: Ack, this is from me waffling on where the validation should live. Apparently I never fully decided

File: osf/models/nodelog.py (hunk 2 above)
Comment (jwalz) [reply osf/models/nodelog.py:148]: I'll update the json files

**Review by brianjgeiger — APPROVED** [review]

**Pull request merged**